pkg/vpcmodel/groupingSelfLoop.go

```diff
@@ -0,0 +1,335 @@
+			candidateLines := groupingSrcOrDst[candidate]
+			// delta between keyLines to candidateLines must be 0
+			mergeGroups := deltaBetweenGroupedConnLines(srcGrouping, keyLines, candidateLines, setsToGroup[key], setsToGroup[candidate])
+			// delta between the keyLines is 0 - merge keyLines
+			if mergeGroups {
+				toMergeCouples = append(toMergeCouples, [2]string{key, candidate})
+			}
+		}
+	}
+	return toMergeCouples
+}
+
+// a group is candidate to be merged only if it has only internal nodes
+func relevantKeysToCompare(groupingSrcOrDst map[string][]*GroupedConnLine) (relevantKeys []string) {
+	relevantKeys = make([]string, 0, len(groupingSrcOrDst))
+	for key, lines := range groupingSrcOrDst {
+		if lines[0].isSrcOrDstExternalNodes() {
+			continue
+		}
+		relevantKeys = append(relevantKeys, key)
+	}
+	return
+}
+
+// optimization to reduce the worst case of finding couples to merge from O(n^4) to O(n^2)
+// where n is the number of nodes.
+// a couple of []*GroupedConnLine is candidate to be merged only if:
+//  1. They are of the same connection
+//  2. If vsis, of the same subnet
+//  3. The src/dst is a singleton contained in the dst/src
+//     in one path on groupingSrcOrDst we prepare a map between each key to the keys that are candidate to be merged with it.
+//     Before the grouping there are at most O(n^20) lines of src -> dst
+//     The last condition implies that each original src -> dst (where src and dst are a single endpoint) can induce a single
+//     candidate (at most), and each singleton key have at most n candidates. Hence, there are at most
+//     O(n^3) merge candidate, which implies O(n^3) time complexity of groupsToBeMerged
+func (g *GroupConnLines) mergeCandidates(groupingSrcOrDst map[string][]*GroupedConnLine, srcGrouping bool,
+	keyToGroupedSets map[string]map[string]struct{}, relevantKeys []string) map[string]map[string]struct{} {
+	// 1. Create buckets for each connection + vsi's subnet if vsi; merge candidates are within each bucket
+	bucketToKeys := make(map[string]map[string]struct{})
+	for _, key := range relevantKeys {
+		lines := groupingSrcOrDst[key]
+		bucket := lines[0].Conn
+		bucket = g.addVsiSubnetToBucket(lines[0].Src, bucket)
+		bucket = g.addVsiSubnetToBucket(lines[0].Dst, bucket)
+		if _, ok := bucketToKeys[bucket]; !ok {
+			bucketToKeys[bucket] = make(map[string]struct{})
+		}
+		bucketToKeys[bucket][key] = struct{}{}
+	}
+
+	keyToMergeCandidates := make(map[string]map[string]struct{})
+	// 2. in each bucket finds for each key the candidates to be merged, in two stages
+	for _, keysInBucket := range bucketToKeys {
+		singletonsInBucket := make(map[string]string)
+		//    2.1 for a group g_1 s.t. the non-grouped src/dst is a singleton,
+		//        all groups in which the grouped dst/src contains the singleton
+		//        2.1.1 finds for each bucket all singletons
+		for key := range keysInBucket {
+			lines := groupingSrcOrDst[key]
+			elemsInkey := elemInKeys(!srcGrouping, *lines[0])
```

Comment:
rename `elemsInkey` to `elemsInKey`

Reply:
> rename `elemsInkey` to `elemsInKey`

These are identical. Perhaps you meant something else?

Reply:
not identical, note the capital `k`

Reply:
Walla